lib/active_merchant/billing/gateways/worldpay.rb

```diff
@@ -245,6 +246,87 @@ def build_authorization_request(money, payment_method, options)
         end
       end
 
+      def add_additional_data(xml, amount, payment_method, options)
+        xml.branchSpecificExtension do
+          xml.tag! 'purchase' do
+            add_level_two_data(xml, amount, payment_method, options)
+            add_level_three_data(xml, amount, payment_method, options)
+          end
+        end
+      end
+
+      def add_level_two_data(xml, amount, payment_method, options)
+        xml.invoiceReferenceNumber options[:invoice_reference_number] if options[:invoice_reference_number]
+        xml.customerReference options[:customer_reference] if options[:customer_reference]
+        xml.cardAcceptorTaxId options[:card_acceptor_tax_id] if options[:card_acceptor_tax_id]
+        xml.tag! 'salesTax' do
+          add_amount(xml, options[:sales_tax][:amount].to_i, options[:sales_tax]) if options[:sales_tax]
+        end
+      end
+
+      def add_level_three_data(xml, amount, payment_method, options)
+        if options[:discount_amount]
+          xml.discountAmount do
+            add_amount(xml, options[:discount_amount][:amount].to_i, options[:discount_amount])
+          end
+        end
+        if options[:shipping_amount]
+          xml.shippingAmount do
+            add_amount(xml, options[:shipping_amount][:amount].to_i, options[:shipping_amount])
+          end
+        end
+        if options[:duty_amount]
+          xml.dutyAmount do
+            add_amount(xml, options[:duty_amount][:amount].to_i, options[:duty_amount])
+          end
+        end
+        if options.include?(:item)
```

Comment:
💬  Comment:

You can write this in just one line =>

```Ruby
if item = options[:item]
```

that will do the two things that you are trying to do